### ISSUE TITLE
(ADR) Average Day Range

### ISSUE THREAD

**Issue by mrjbq7** [TA-Lib/ta-lib#367]
> [!IMPORTANT]
> **RULED: default period 14.** `min = 1`. ADR% is a separate card (proposal-drafts#69), not a parameter here.
>
> The `Open questions` section further down is **historical** — it records what was open
> when the card was written, not what is. The ruling comment on this issue is the current
> state and wins over the body wherever they disagree.

A request was made for the "Average Day Range" indicator

https://github.com/TA-Lib/ta-lib-python/issues/575

---

## Proposal

**SHIP.** `TA_ADR` = `SMA(high - low, n)`, one output, group `Volatility Indicators`,
default period 14. No blocker: an external golden already exists offline and is
bit-identical to the proposed body (MEASURED, below).

Two scoping decisions, both argued in Formula:

- **Form A** (mean of the ranges), not form B (`SMA(high,n) - SMA(low,n)`). Sources split
  2-2; form A wins on conditioning, MEASURED.
- **No `ADR%` output in this function.** The two published percentage forms differ by up to
  **22.13 %** on the reference corpus (MEASURED) — a variant choice, not a rounding. Separate
  card.

**Not what the requester asked for, and the card must say so.** TA-Lib/ta-lib-python#575 (state CLOSED,
2024-01-20) pins no formula: the requester offered "the strategy.py-file of freqtrade" and
never posted it; mrjbq7 redirected to `TA-Lib/ta-lib`, where no follow-up exists
(`gh search issues --repo TA-Lib/ta-lib` for both `ADR` and `average day range` → `[]`). The
only freqtrade code naming "average day range" is the Solipsis/Schism family, and it is
`close.rolling(24).max() - close.rolling(24).min()` — a Donchian-style channel width on
*close*, no average, no high/low, already expressible as
`TA_SUB(TA_MAX(close,24), TA_MIN(close,24))`. `TA_ADR` answers the *name* and the demand
behind it, not that code.

**Why a function and not a documented recipe** (the TA-Lib/ta-lib-proposal-drafts#49/#51/#56 question). TA-Lib/ta-lib-proposal-drafts#49 MAD is
`AVGDEV` byte-for-byte, so a doc alias is the whole fix; TA-Lib/ta-lib-proposal-drafts#51 BBW is arithmetic on an existing
function's *output*. ADR needs a second array materialised (`TA_SUB(high, low)`) before any
shipped function can consume it. The in-tree precedent is exact: `TA_QSTICK` = `SMA(close −
open, n)` ships as a first-class function with this same composite differential
(`src/tools/ta_regtest/ta_test_func/test_composite1.c:2279-2340`). ADR is that function with
`high − low` in place of `close − open`.

## What it is

The arithmetic mean of the last `n` bar ranges, `high - low`. Same family as ATR, and
deliberately the narrower one: the range excludes the overnight gap, so ADR answers "how far
does price travel *within* a bar" where ATR answers "how far including the jump into the
bar". TradingView's own page says it in those words — ADR% "does not take gaps into account".

No inventor and no book. The authoritative published definitions are commercial-platform
documentation (TradingView support, TC2000 PCF help), both FETCHED and quoted under
References, plus one runnable open-source implementation (kand).

## Formula

Let `n = optInTimePeriod`.

**Form A (proposed):**

```
Range_i = high[i] - low[i]
ADR_i   = ( SUM_{k=i-n+1..i} Range_k ) / n
```

C body — `ta_codegen/input/sma/sma.c:38-90` with `inReal[x]` replaced by
`(inHigh[x] - inLow[x])` and nothing else changed, exactly as
`ta_codegen/input/qstick/qstick.c` did for `close - open`:

```c
lookbackTotal = (size_t)(optInTimePeriod - 1);
if( startIdx < lookbackTotal ) startIdx = lookbackTotal;
if( startIdx > endIdx ) { *outBegIdx = 0; *outNBElement = 0; return TA_SUCCESS; }

periodTotal = 0.0;
trailingIdx = startIdx - lookbackTotal;
i = trailingIdx;
if( optInTimePeriod > 1 )
   while( i < startIdx ) { periodTotal += (double)(inHigh[i] - inLow[i]); i = i + 1; }

outIdx = 0;
while( i <= endIdx )
{
   periodTotal += (double)(inHigh[i] - inLow[i]);
   i = i + 1;
   tempReal     = periodTotal;
   periodTotal -= (double)(inHigh[trailingIdx] - inLow[trailingIdx]);
   trailingIdx  = trailingIdx + 1;
   outReal[outIdx] = tempReal / (double)optInTimePeriod;   /* divide, NOT * (1/n) */
   outIdx = outIdx + 1;
}
*outNBElement = outIdx; *outBegIdx = startIdx;
```

`lookback = optInTimePeriod - 1`. The trailing read precedes the store, so `outReal` may
alias `inHigh` or `inLow`.

**No NaN/Inf can escape (#112), on both paths.** The only division is by the integer period,
never zero. The only route to ±Inf is accumulator overflow, and `periodTotal` sums at most
`n × (TA_REAL_MAX − TA_REAL_MIN)` = 1e5 × 6e37 = 6e42 (`include/ta_defs.h:115-116`), far
under `DBL_MAX` — the identical exposure `TA_SMA` already ships with.

**Rejected variants:**

| Variant | Expression | Why not |
|---|---|---|
| **Form B — difference of averages** (both TradingView pages) | `SMA(high,n) - SMA(low,n)` | Algebraically identical, numerically worse: it subtracts two ~90-magnitude means to get a ~3-magnitude answer, so the result inherits the larger scale's rounding. MEASURED on the SREF 252-bar corpus against exact rational arithmetic — form A is the correctly-rounded result on **every** bar at n ≤ 50 (0 ULP) and within 5 ULP at n = 100; form B is correctly rounded on 6/233 bars at n = 20 and is off by up to **620 ULP** at n = 50. The two also differ bitwise on 227 of 233 bars at n = 20 (max rel 2.19e-14), so the spelling has to be pinned either way. |
| **Form C — Qullamaggie ADR%** | `100 · ( SMA(high/low, 20) − 1 )` | Mean of *ratios*; different series, different unit. |
| **Form D — TradingView ADR%** | `( SMA(high,14) − SMA(low,14) ) / close · 100` | Ratio of *means*. MEASURED against form C at n=20 on SREF: relative difference (denominator `|C|`) max **22.13 %**, min **0.029 %**. Two indicators sharing a name. |
| A per-calendar-day ADR | average of daily H−L regardless of chart timeframe | Needs a session/calendar concept. TA-Lib has none. See Notes. |

Source split on A vs B is **2-2**, not one outlier: TC2000 and kand spell A, both TradingView
pages spell B. The decision is made on conditioning plus the in-tree QSTICK precedent, not on
a vote.

## Proposed API

```c
TA_RetCode TA_ADR( int    startIdx,
                   int    endIdx,
                   const double inHigh[],
                   const double inLow[],
                   int    optInTimePeriod,   /* From 1 to 100000, default 14 */
                   int   *outBegIdx,
                   int   *outNBElement,
                   double outReal[] );

int TA_ADR_Lookback( int optInTimePeriod );   /* = optInTimePeriod - 1 */
```

`ta_codegen/input/adr/adr.yaml`:

- `group: Volatility Indicators` — existing group 3, alongside ATR/NATR/TRANGE
  (`src/ta_abstract/ta_group_idx.c:288-292`). No new `TA_GroupId`.
- `inputs: inPriceHL`, `price_components: [high, low]` — the bundle DONCHIAN/MIDPRICE/MEDPRICE
  already use (`ta_codegen/input/donchian/donchian.yaml:5-8`).
- `optInTimePeriod`: `range: [1, 100000]`, `default: 14`, `suggested: [1, 200, 1]`,
  `display_name: Time Period`, `hint: Time period` — the ATR row verbatim
  (`ta_codegen/input/atr/atr.yaml:10-16`).
- one output `outReal`, `flags: [line]`; function `flags: [stream]`.

**Default 14, sourced.** One FETCHED vendor page (TradingView ADR%/ATR%: `ta.sma(high,14) -
ta.sma(low,14)`) plus in-library ATR/NATR parity (`atr.yaml:12`, `natr.yaml:14`). That is the
whole case — kand casts no vote (it declares no default; the period is a required argument,
and the `14` in `adr::lookback(14)` is an illustration in a doctest whose sibling
`adr()` doctest and unit test both use 3). Competing values to record in `adr.md` and not
follow: TradingView's own ADR *indicator* page works its example at **7**, TC2000's examples
at **10**, the Qullamaggie screener convention at **20**.

**`min = 1`, not kand's 2.** `ADR(1)` is exactly `high[i] - low[i]`, a defined and useful
series (`TA_SUB(high, low)`), and `SMA`/`QSTICK` both admit 1 (`sma.yaml:13`,
`qstick.yaml:14`). MIDPRICE and DONCHIAN start at 2 because their window extremum is
degenerate at 1; ADR's is not. Do **not** set `period1_identity`: that flag promises the
output is a *bit-exact copy of an input value* (`ta_codegen/generator/tests/period1_suite.rs:1-8`),
and `high[i]-low[i]` is not a copy of anything. `qstick.yaml` is structurally identical,
likewise unflagged, and ships green.

Flags **not** set: `overlap` (a range is not a price level); `unstable_period` (finite window,
no infinite recurrence — so none of the seven hand edits a `TA_FUNC_UNST_*` id costs);
`TA_FUNC_FLG_START_DEP` (#127 — bar `i` depends only on `high/low[i-n+1 .. i]`).

**Required deliverable: `ta_codegen/input/adr/adr.md`.** Every input dir ships one
(`qstick/qstick.md`, `donchian/donchian.md`, `rma/rma.md`); it is the source of the Rustdoc
prose and the ta-lib.org function page. It must carry: the range-vs-true-range distinction,
the "the bars you pass are the days" sentence (see Notes), the competing defaults above, the
`high < low` statement, and the difference from the TA-Lib/ta-lib-python#575 requester's freqtrade code. The
runnable doctest is **not** written here — the generator synthesises it from the input bundle
(`ta_codegen/generator/src/backends/rust_doc.rs:634`; see the emitted shape at
`ta_codegen/output/rust/library/src/ta_func/qstick.rs:226-243`).

## References

**FETCHED for this card:**

- **TradingView — "How are ADR% and ATR% calculated?"** —
  https://www.tradingview.com/support/solutions/43000734653-how-are-adr-and-atr-calculated/
  Form B verbatim as Pine on the chart's own bars: `smaHigh = ta.sma(high, 14)`,
  `smaLow = ta.sma(low, 14)`, `adr = smaHigh - smaLow`, `adrp = adr / close * 100`; and
  "ADR% focuses on daily movements and does not take gaps into account."
- **TradingView — "Average Daily Range (ADR) indicator"** —
  https://www.tradingview.com/support/solutions/43000695003-average-daily-range-adr-indicator/
  "the average of the high values over the past seven days, then the average of the low values
  for the same period ... the difference between these values". Inputs are **Length** ("the
  number of days used") and **Timeframe** — the "day" is a user input, not a session.
- **TC2000 PCF help — Average Daily Range (ADR)** —
  https://help.tc2000.com/m/69445/l/1993818-average-daily-range-adr
  Form A verbatim: `AVG(Hz-Lz, x)`; examples `AVG(H-L, 10)`, `XAVG(H-L, 10)`, `AVG(H2-L2, 10)`.
  No percentage form.
- **kand `kand/src/ta/ohlcv/adr.rs`** (repo `kand-ta/kand`, read via the contents API) —
  `Daily Range = High - Low; ADR = SMA(Daily Range, period)` (`:31-32`),
  `lookback = period - 1` (`:20-22`), `adr()` delegating to `sma::sma` over a materialised
  `ranges` vector (`:93-101`), an incremental `adr_inc` (`:135-165`), and the golden test
  vector used below (`:169-342`). Confirms `to_be_deleted/kand-catalog.md:242` against source.
- **werkkrew/freqtrade-strategies `Solipsis_v5.py:190-193`** —
  `informative['adr'] = informative['1d-high'] - informative['1d-low']` over
  `close.rolling(24).max()` / `.min()`. Not ADR.
- **`TA-Lib/ta-lib-python#575`** — full thread; the requester's implementation was offered but
  never posted.

**Cited, not fetched:** LuxAlgo's ADR blog post ("a 14-day period is often used"); the
Qullamaggie ADR% convention `100·(SMA(H/L,20)−1)`, which is community-standard on TradingView
and AmiBroker forums with no vendor documentation page behind it. The freqtrade `Schism`
variant is not independently verified;
only `Solipsis_v5.py` was read locally.

## Sample input/output (board policy)

Real numbers from an implementation outside this project: kand's checked-in unit-test vector,
`kand/src/ta/ohlcv/adr.rs:169-342` — 100 high/low bars with 98 expected ADR values at
period 3, printed in the file to 20 significant digits.

MEASURED: the proposed body above reproduces **all 98 values bit-identically** (98/98
byte-equal doubles; relative difference 0.0 on every bar). Against exact rational arithmetic
both are within **7.4e-17** relative — i.e. correctly rounded on every bar.

Self-contained excerpt (the first 12 bars; at n=3 each output depends on a 3-bar window, so
these 10 outputs are kand's first 10 unchanged):

```
inHigh = {35266.0,35247.5,35235.7,35190.8,35182.0,35258.0,35262.9,35281.5,35256.0,35210.0,35185.4,35230.0}
inLow  = {35216.1,35206.5,35180.0,35130.7,35153.6,35174.7,35202.6,35203.5,35175.0,35166.0,35170.9,35154.1}

TA_ADR(0, 11, inHigh, inLow, 3, ...)
outBegIdx = 2   outNBElement = 10
outReal   = {48.866666666666184, 52.266666666667639, 48.066666666668119,
             57.266666666670062, 57.333333333335759, 73.866666666668607,
             73.100000000000975, 67.666666666666671, 46.5,
             44.800000000000487}
```

For orientation on the corpus the differential will run on
(`TA_SREF_{high,low}_daily_ref_0_PRIV`, `src/tools/ta_regtest/test_data.c:38,:65`, bound at
`ta_regtest.c:611-613`): at n=14, `outBegIdx = 13`, `outNBElement = 239`, first eight values
MEASURED `2.9442857143, 3.0917857143, 3.0042857143, 3.1739285714, 3.1157142857, 3.0825000000,
3.1739285714, 3.1628571429`.

Note `gDataHigh`/`gDataLow` are a *different*, 10000-bar array (`ta_gDataHigh.c:1`) reached
only by `test_abstract.c:109-112` and the SUPERTREND/KC/RMA suites. Do not source a
verification number from it for a `TA_History`-driven test.

## Oracles

| Oracle | Covers ADR | Arm status | Work to add the arm |
|---|---|---|---|
| **kand** (`kand-ta/kand`, Rust) | **yes** — `ohlcv/adr.rs`, form A | **frozen vector, no wiring** | Transcribe the 100-bar input and 98 expected values as `static const double` beside the QSTICK Tulip vector (`test_composite1.c:2261-2270`). No network, no capture binary, no 5th server. |
| `ta_pandas_serve` | **no** `adr` module (package-wide grep; `volatility/` = aberration, accbands, atr, bbands, ce, cvi, donchian, hvol, hwc, kc, massi, natr, pdist, rvi, thermo, true_range, ui) | *composable* | ~5 lines: `compute_ADR` = `ta.sma(pd.Series(high) - pd.Series(low), length=n, talib=False)`. |
| `ta_tulip_serve` | **no** — no `adr.c` in the 109-file `indicators/` listing | n/a | — |
| `trading_signals_serve` | **no** — 8.3.0 `dist/volatility` is ABANDS…UI, no ADR | n/a | — |
| `ta4j_serve` | **no** — both ta4j-core jars (Maven Central `org.ta4j:ta4j-core:0.22.6` **and** `:0.24.1`) grepped for `adr`/`DayRange`/`DailyRange`: nothing | n/a | — |

**The board's publishable bar is met today.** kand is a runnable outside implementation of
form A and its own published output is bit-identical to ours. TC2000 and TradingView are
published statements of forms A and B.

**Two shape mismatches** if anyone later wires kand live rather than freezing the vector:
`adr()` NaN-pads the first `period-1` outputs (`adr.rs:39`), and returns
`KandError::InsufficientData` when `len <= lookback` (`adr.rs:76`) where TA-Lib returns
`TA_SUCCESS` with `outNBElement == 0`. Both must be normalised or the arm reports false
mismatches on exactly the edge cases worth testing. The frozen-vector arm has neither problem.

**The pandas arm is circular unless pinned.** `pandas_ta_classic/overlap/sma.py:26-31` routes
to TA-Lib's own `SMA` when `talib=True`; pass `talib=False` explicitly. On the default path it
is stronger than "weak" implies — MEASURED against the TA-Lib SMA recurrence over the SREF
range series: **bit-identical** at n = 2, 5, 14, 20, 50; max rel **7.5e-16** at n = 100 and
**6.5e-16** at n = 200. The logical caveat still stands: it is our decomposition with their
mean, so it confirms window alignment and lookback, not that ADR *is* `SMA(H−L)`.

**kand's own SMA order is not measured here.** `adr.rs:101` delegates to `sma::sma`, and
`kand/src/ta/ohlcv/sma.rs` was not read for this card. Nothing rests on it: the arm is the
n=3 vector, which is bit-identical. A live kand arm at other periods would need the 1e-12
band rather than `memcmp`.

## Verification plan

1. **Composite differential, bit-exact.** Append an ADR leg to `test_composite1.c` beside the
   QSTICK legs (`:2279`) — the file is already registered in both `CMakeLists.txt:660` and
   `src/tools/ta_regtest/Makefile.am:50`, so no source-list churn. Reference =
   `TA_SUB(high, low)` → `TA_SMA(that, n)`, compared with `memcmp`. Bit-exact by construction:
   `TA_SUB` stores exactly `inHigh[i]-inLow[i]` (`ta_codegen/input/sub/sub.c:29-31`, lookback 0)
   and the fused body accumulates the same doubles in the same order.
   **Feed `TA_SMA` the SAME `startIdx` as `TA_ADR`**, per `test_composite1.c:2313-2316` — "so
   the lookback clamp is compared too, not just the values". ADR's lookback is `n-1`, identical
   to SMA's, with identical seeding, so both anchors reach the same window by the same
   add/subtract order. (The DPO card's opposite instruction exists only because DPO's lookback
   is `max(n-1, n/2+1)` ≠ SMA's.)
   Grid `n ∈ {1,2,3,5,14,20,50,100}` × the `qstickStartGrid` starts.
2. **External golden.** Freeze kand's 100-bar vector and 98 expected values, `tol 1e-12` /
   `abs 1e-14` (QSTICK's band, `test_composite1.c:2275-2276`) — a safety band, not a necessity;
   the values are bit-identical MEASURED. Without this leg nothing in the tree catches "ADR is
   not `SMA(H−L)`", because both sides of leg 1 share that assumption
   (`to_be_deleted/composite-functions.md` §4).
   *Decision recorded:* freeze the kand vector **and** add the 5-line pandas arm. The vector is
   the formula oracle; pandas is an independently-written rolling mean that keeps the window
   alignment honest across the whole SREF grid rather than at one period.
3. **Free coverage**, automatic once the YAML lands, in strength order:
   `--xlang-hash` / `server_verify` at zero tolerance across C/Rust/Java/.NET (this is the
   cross-language *value* gate); `--fuzz-064`; the #94 param-grid boundary sweep under
   ASan/UBSan; `checkInPlaceAliasCorrect` (#130, `test_abstract.c:3069`), which runs every
   (real input component × real output) pair aliased vs separate and compares bitwise —
   `inPriceHL` gives it two pairs; and, with `[stream]`, `stream_verify`'s
   Open/Update/Peek/OpenAndFill per-bar compare.
   `--codegen --function=ADR` contributes the API/metadata legs **only**: it value-compares
   nothing for a post-cutover function, prints `NO VALUE COMPARISON`
   (`test_codegen.c:10091-10098`) and exits 0. #108's output-distinctness guard is emitted only
   for `outputs.len() >= 2` (`backends/c.rs:973`), so it covers ADR not at all.
4. **Hand-written edge cases** (the ones no generic gate reaches):
   - `n == 1` — assert `outReal[k] == inHigh[beg+k] - inLow[beg+k]` exactly, `outBegIdx == startIdx`.
   - `n == 2` — smallest seeded case; assert `outBegIdx` explicitly.
   - `startIdx == endIdx == lookback` — exactly one output.
   - `nbBars < lookback` — `TA_SUCCESS`, `outNBElement == 0`, `outBegIdx == 0`.
   - `high[i] == low[i]` throughout — assert exactly `0.0`, not near-zero.
   - `high[i] < low[i]` — assert the output is the negative mean and `TA_SUCCESS` (see Notes).
   `outReal` aliasing `inHigh`/`inLow` needs **no** hand-written case: it is exactly what the
   #130 gate does automatically for this input bundle.

## Open questions

1. **Default period: 14, 20, 10 or 7?** 14 buys ATR/NATR parity inside this library and is
   TradingView's documented ADR% period; 20 is what the largest identifiable user community
   (ADR% screener users) means by "ADR"; TC2000's examples use 10; TradingView's own ADR
   indicator page works its example at 7. Only one can be the default. Recommendation: 14.
2. **`min = 1` or `2`?** Recommendation above is 1 (SMA/QSTICK parity, and `ADR(1)` is a
   defined series). kand, MIDPRICE and DONCHIAN all start at 2. This is a house-convention
   judgement, not a source claim — flagged so it is decided rather than inherited.
3. **Ship an `ADR%` at all, and which form?** C and D differ by up to 22.13 % MEASURED and both
   have real user bases. The structural precedent is cheap once the variant is picked — `NATR`
   is `100·ATR/close` with an exact `close != 0.0` guard
   (`ta_codegen/input/natr/natr.c:190-199`). Recommendation: separate card.
4. **What `adr.md` says about TA-Lib/ta-lib-python#575.** The issue is closed and there is nothing to close; the
   live decision is whether the function page names the requester's `MAX(close,24) −
   MIN(close,24)` construction as a different thing available via `TA_MAX`/`TA_MIN`/`TA_SUB`,
   or stays silent. Recommendation: name it — the alternative is a page that looks like it
   answered a request it did not.

## Notes

- **The "Day" in the name is not a session.** If "daily high minus daily low" meant "resample
  to calendar days regardless of chart timeframe", ADR would need a session/anchor concept
  TA-Lib lacks and this card would be blocked. It does not: TradingView's ADR indicator exposes
  **Timeframe** as a user input, and TC2000's PCF is `AVG(H-L, x)` on whatever bars the chart
  holds. The bars the caller passes *are* the days — the convention VWAP shipped under
  (`ta_codegen/input/vwap/vwap.md`). Reuse that wording; never let the page say "daily"
  unqualified.
- **`high < low` is not an error.** Garbage input yields a negative mean and `TA_SUCCESS`; no
  input validation exists or should be added (the library validates ranges and parameters, not
  price sanity). Say so on the function page, since the output is documented as a range.
- **Not ATR, and say so.** `TA_TRANGE` is `max(H−L, |H−Cprev|, |L−Cprev|)`; ADR uses only the
  first term, so on a gapping instrument ADR is systematically smaller. That is the point of
  having both.
- **Not Donchian width either** — `MAX(high,n) − MIN(low,n)` is the window extreme, ADR the
  mean of per-bar ranges; `TA_DONCHIAN` already ships the former.
- **Streaming tier: ordinary, and the emitter already handles the exact shape.** For the
  identical two-array trailing read, `src/ta_func/ta_QSTICK.c:256,274-281` carries a **single**
  `double *ring_trailingIdx_derived` holding `(inClose - inOpen)` — one ring, not two — plus a
  running total. `flags: [stream]` from the first commit, not a follow-up. kand ships an
  `adr_inc` with the same state (`adr.rs:135-165`).
- **Registration and ABI.** `extern const TA_FuncDef TA_DEF_ADR;` plus two entries in the
  generated `src/ta_abstract/ta_group_idx.c` (declaration block and `TA_PerGroupFunc_3`).
  There is no `TA_FuncID` in this tree. ABI impact is new exported symbols only — `TA_ADR`,
  `TA_S_ADR`, `TA_ADR_Lookback` and the streaming family (`TA_ADR_Open`/`Update`/`Peek`/
  `Close`/`OpenAndFill`/`UpdateAndFill`/`Value`/`Clone`, per `include/ta_func.h:11195-11237`).
  Nothing is renumbered.
- **Alphabetical position is free in both directions.** `ADR` sorts between `ADOSC` and `ADX`
  globally (`ta_group_idx.c:46-49`) and **first** within Volatility Indicators (ahead of `ATR`,
  `:289`). Neither costs anything: the `ta_ref_serve` hazard from AC (#228) was the global
  `list_functions` JSON payload, where only the alphabetically first entry opens the array
  without a leading comma, and it was fixed generically —
  `scripts/serve_version.py:170-181` re-normalizes the comma across the surviving entries
  "rather than assumed".
- **No naming collision** — `grep -rn '\bADR\b'` over `src/ta_abstract/`, `ta_codegen/input/`
  and `include/` → empty. 182 directories under `ta_codegen/input/` (181 functions + `helpers`), none of them `adr`.
- **Neighbouring open cards, all in the volatility family:** TA-Lib/ta-lib#358 CVI (`ROC(EMA(high−low, n), r)`
  — same `high−low` intermediate; sequence with ADR and share whatever accumulation CVI
  settles on), TA-Lib/ta-lib-proposal-drafts#44 HVOL, TA-Lib/ta-lib-proposal-drafts#58 REI, TA-Lib/ta-lib#366 RVI.
- **Process gotchas.** Run a *full* `cargo run -- generate` (`--func=ADR` skips `Core.java` and
  the other whole-corpus files); run `scripts/build.py servers`, not `build`, or the
  new-function gate coverage is silently stale; add the `CHANGELOG.md` entry, forgotten for
  CMOU.

**Comment by mario4tier** [comment]
## Ruling: default period 14. Moved to Todo.

`optInTimePeriod` defaults to **14** — ATR/NATR parity inside this library, and TradingView's
documented ADR% period.

The card's other two recommendations stand and were not disputed: `min = 1` (SMA/QSTICK parity;
ADR(1) is a defined series), and ADR% gets its **own card** rather than a parameter here — that is
TA-Lib/ta-lib-proposal-drafts#69, already open and held on its own form C-vs-D ruling.

`adr.md` names the requester's `MAX(close,24) - MIN(close,24)` construction from ta-lib-python#575 as a
different thing reachable via `TA_MAX`/`TA_MIN`/`TA_SUB`, per the card's recommendation.

**Comment by mario4tier** [comment]
Taking this one, as part of batch **w1a** (#367 ADR, #364 PVT, #370 RVOL) on branch `feat/w1a-adr-pvt-rvol`, landing together on `dev`.

Decisions applied, stated here so they are visible before the code lands:
- `hint: Average Day Range` (matches the card title and the CHANGELOG line). "Average Daily Range" goes in `adr.md`'s `## Aliases`, which feeds the rustdoc `#[doc(alias)]` and site search.
- Form A (mean of per-bar ranges), default 14, range `[1, 100000]`, no `period1_identity`.
- Golden values: kand's LICENSE gets checked before anything is transcribed. If it is not MIT/Apache-2.0 I follow the in-tree precedent instead of copying its arrays.

**Comment by mario4tier** [comment]
Shipped on `dev`: 98c451d27

Form A (mean of per-bar ranges), default 14, range `[1, 100000]`, no `period1_identity`. Golden values from kand 0.2.2, reproducing all 98 of its published doubles bit-for-bit at zero tolerance.
